internal/stacks/stackruntime/internal/stackeval/terraform_hook.go

```diff
@@ -81,13 +84,23 @@ func (h *componentInstanceTerraformHook) PreApply(addr addrs.AbsResourceInstance
 	if h.resourceInstanceObjectApplyAction.Len() == 0 {
 		h.resourceInstanceObjectApplyAction = addrs.MakeMap[addrs.AbsResourceInstanceObject, plans.Action]()
 	}
-	h.resourceInstanceObjectApplyAction.Put(
-		addrs.AbsResourceInstanceObject{
-			ResourceInstance: addr,
-			DeposedKey:       dk,
-		},
-		action,
-	)
+	localObjAddr := addrs.AbsResourceInstanceObject{
+		ResourceInstance: addr,
+		DeposedKey:       dk,
+	}
+
+	// We may have stored a previous action for this resource instance if it is
+	// planned as create-then-destroy or destroy-then-create. For those two
+	// cases we need to synthesize the compound action so that it is reported
+	// correctly at the end of the apply process.
+	if prevAction, ok := h.resourceInstanceObjectApplyAction.GetOk(localObjAddr); ok {
+		if prevAction == plans.Delete && action == plans.Create {
+			action = plans.DeleteThenCreate
+		} else if prevAction == plans.Create && action == plans.Delete {
+			action = plans.CreateThenDelete
+		}
+	}
```

Comment:
📝

The seemingly arbitrary nature of this reconstruction of the compound planned action is unsettling, but I'm not sure there's a better option.

One that I considered and so far rejected on aesthetic grounds is storing a map of resource address to slice of actions, and updating the `ResourceInstanceObjectAppliedAction` method to return zero or more actions, then tracing that change through its caller. This doesn't seem clearly better to me and is a more invasive change, but I'm open to other opinions (and even more so to better ideas).

Reply:
This is indeed quite unfortunate, particularly because in other parts of the system (mimicking what we'd previously done with the JSON plan output) we intentionally chose to always present these compound action verbs as pair of verbs -- `["create","delete"]` and `["delete","create"]` -- to try to achieve outward-facing consistency in the other direction, knowing that Terraform Core would be splitting them into two distinct actions during apply anyway. :confounded: 

If the goal is to present the impression that each resource instance object has only one apply phase that spans across both operations then I agree this seems like the most straightforward way to achieve that, despite it being a little abstraction-inversion-ish. The other way we could go, that would be a bit more honest about what's going on, is for these objects to go through the (applying, applied) sequence twice with different actions each time, but I know the Terraform Cloud UI for Stacks is not really built for that to come across clearly, and so I defer to your judgement on what's the most appropriate way for Terraform Core to describe what it's doing here.

Reply:
The goal isn't to present a single apply phase for a resource instance when replacing—we are still emitting two distinct applying/applied events to the stream.

The specific issue that this change is fixing is only the change summary count emitted by the stacks runtime for each component. What we're trying to achieve is consistency between plan and apply.

Historically we've always counted a replace operation in the change summary counts as "1 add, 1 remove". The structure of the code for the stacks runtime for generating the apply change summary counts depends on the hook's map of addr to action. Previously any replace operation would overwrite the first action of the pair with the second, resulting in a change summary of either "1 add" or "1 remove" (depending on create-before-destroy).

If you think it would be clearer to adjust the structure of the code to emit a slice of actions as I described above, I'd be happy to make that change. What I'm trying to avoid here is leaving a mess of confusion for the next person.

Reply:
Here's half of what I mean by the above, with no test coverage for the caller of this hook: e00f62738f67f1ef2184d8910bd788375a1ec443

Reply:
I'm honestly not really sure what to prefer here. This inconsistency has caused similar problems before, and so our much earlier decision to pretend in the machine readable formats that these actions are _always_ decomposed into their primitive actions was made with that experience in mind, but I'm not really sure that it actually worked, as opposed to just moving the inconsistency somewhere else.

An advantage of hiding the fact that the two "replace" actions exist as single actions is that in theory it allows callers to be written "naively" by just counting up how many times they see "create" and "delete" while letting Terraform Core worry about how they interact with one another to produce the compound actions.

However, it's looking like a clear disadvantage is how hard it is to consistently present that abstraction at every "edge" where we present data for external consumption, leading to consumers having to work around Terraform Core's modelling mistakes. It would perhaps be easier for the whole system to just be aware that "create before destroy" and "destroy before create" are actions and decompose them only at the edges of the system where it's actually important decompose them.

I don't see a clear answer here, and so I would suggest that we go with what you did and try to get some experience with it between now and general availability, and revisit this only if we find that it's causing some kind of maintenance headache elsewhere.